blog: Predefined models exp

flippedcoder opened the pull request:
This is a post about how you can use DVC experiments to help fine-tune a pre-trained model like AlexNet or SqueezeNet.

jorgeorpinel commented:
Cool! Please let us know when you'd like reviewers involved. Thanks

jendefig commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,148 @@
+Since the project has everything we need to initialize the models, we can start
+training and comparing the differences between them with the ants/bees dataset.
+Running experiments to get the best tuning for each model can make it difficult
+to see which changes led to a better result. That's why will be using DVC to
```
insert "we"

shcheklein replied:
Last two sentences with minor modifications could make it to the `descpritionLong`, which will benefit from it a lot I think.

jendefig commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,148 @@
+Let's break down the different parts of this command.
+
+- `--name` is the title of the stage
+- `--deps` defines the files and directories that a stage needs in order to run
```
maybe add: "short for dependencies, defines the ..."  This is probably is clear to the software engineering set, but I think will help clarify for those that are more green.

dberenbaum reviewed (commented):
Fun article idea!

With the hyperparameter tuning post, it seemed mostly useful to highlight a dvc pattern that wasn't documented elsewhere - how to do hyperparameter searches using dvc experiments. This article keeps to the basic experiments workflow that users can already learn from other docs, so it seems like the value is more about the interest in transfer learning. In that case, interesting results might be more important here, or at least a meaningful approach to how/why to try different experiments. It might make sense to touch base with @daavoo whose background is in computer vision.

Also, could you add attribution to the original source code used to derive this example (and where to get the data)? I see something sort of similar in https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html. One idea for this post would be to update the code to make it more dvc-like and then show how this can make it easier to track different experiments. Some examples:

* Adding `model_name` to the parameters makes it so you can see at a glance whether an experiment is alexnet or squeezenet.
* Logging each epoch using dvclive or using checkpoints enables tracking results per epoch, which currently are getting printed to the screen and then lost.
* Saving both the training and validation metrics shows a more complete picture (it looks like only the training metrics are being saved now) and you can choose to view one or the other or both.

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,352 @@
+title: Using Experiments to Improve Pre-trained Models
+date: 2021-08-24
+description: |
+There are times it will be easier to take a pretrained model and fine-tune it to work with your data. You can do that with DVC experiments.
```
💅 style, make it < 80

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,352 @@
+There are times it will be easier to take a pretrained model and fine-tune it to work with your data. You can do that with DVC experiments.
+descriptionLong: |
+When you're working with a pretrained model and you want to test different fine-tuning values, using DVC experiments can help you do that faster.
+picture: 2021-08-17/pretrained-models.png
```
2021-08-17 - fix the date?

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,352 @@
+[virtual environment](https://python.readthedocs.io/en/stable/library/venv.html)
+and install the dependencies with:
+
+```bash
````
bash -> dvc

shcheklein commented:
<img width="978" alt="Screen Shot 2021-08-05 at 5 52 59 PM" src="https://user-images.githubusercontent.com/3659196/128439529-d71df488-4f4e-45cd-93f0-1879fdcd10fa.png">

Something is wrong, does it work for you locally?

https://dvc-org-pre-defined-mod-kedssi.herokuapp.com/blog/pre-defined-models-exp is also broken it looks like ..

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,352 @@
+or even people. This is called transfer learning and it can save a lot of time
+on developing a model from scratch.
+
+For us to take advantage of transfer learning, we can use fine-tuning to adopt
```
Should we use this in the title/descriptions? Something "Faster Transfer Learning with Experiments"? (can capture more attention?)

or `Using Experiments for Transfer Learning`

shcheklein replied:
Smooth Transfer Learning

or 

Pain-free Transfer Learning :)

flippedcoder replied:
I like `Using Experiments for Transfer Learning`!

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,352 @@
+
+There are plenty of machine learning models available that have been trained to
+solve one problem and the knowledge gained from that can be applied to a new,
+yet related problem. For example, a model like AlexNet has been trained on
```
good to have links - to transfer learning, AlexNet

flippedcoder replied:
I have some links for AlexNet and SqueezeNet a little later and I added a link explaining transfer learning.

shcheklein replied:
yep, it's fine .. usual convention (for docs) is to put link the first time you mention the concept. Not a big deal clearly.

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,352 @@
+
+## Initialize the pre-trained model
+
+We'll be fine-tuning the AlexNet model and the SqueezeNet model to classify
```
is there some image we can put - with bees and ants :)

flippedcoder replied:
That's a good idea! I'll run it by Jeny. :)

shcheklein replied:
I meant something actual, from the dataset?

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,352 @@
+```git
+$ git status
+$ git add .
+$ git commit -m "added DVC pipeline"
````
this not a DVC pipeline yet ... it should be "DVC init"

shcheklein replied:
is it resolved? should we update the git commit message?

flippedcoder replied:
I removed this section because we all agreed that we don't need to set up DVC in every blog post.

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,352 @@
+to see which changes led to a better result. That's why will be using DVC to
+track changes in the code and the data.
+
+## Add a DVC pipeline
```
It doesn't look about pipelines, unless I missing something

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,352 @@
+$ git commit -m "added DVC pipeline"
+```
+
+With DVC in place, we need to add a stage to the pipeline so that we can train
````
`Put DVC in place` - good name for the section :)

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,352 @@
+We'll add a stage for the training step in this example. To do that, run:
+
+```dvc
+$ dvc stage add --name train --deps data/hymenoptera_data --deps pretrained_model_tuner.py --metrics-no-cache --params lr,momentum results.json python pretrained_model_tuner.py
````
so long lines could be split with `\` - please take a look at the docs, there are examples in the pipelines get started

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,352 @@
+This will generate a new stage in the `dvc.yaml` file. If you take a look at it,
+you'll see something like:
+
+```yaml
````
We can start with this YAML (and then briefly explain the helper command that we were using to generate it). Also it would be good to explain why do we need this file (for DVC to start managing data, understand where metrics to pay attention to, etc - some friendly explanations)

flippedcoder replied:
I have the description of the `dvc stage add` command before this YAML. So you want me to move that description below the YAML? And I added friendlier explanations for each part of the command.

shcheklein replied:
yep, it can be simler to some extent - YAML is more or less friendly and something folks are already familiar vs some new command they haven't seen before. No strong opinion though. @dberenbaum wdyt? :)

dberenbaum replied:
Yes, I think the first step could be cloning the repo and then explaining parts of the repo as necessary like `dvc.yaml`. I wouldn't repeat the steps that you've already done in that repo since it could be confusing and it takes longer to get to the transfer learning part. You can also give people the option to skip the explanations and jump right into the transfer learning part. I'm not even sure we really need a detailed explanation of the setup at the start of each blog post, but at least we should give the option of getting right to the main subject matter.

flippedcoder commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,352 @@
+images of bees and ants. You can find the project we're working with in
+[this repo](https://github.com/iterative/pretrained-model-demo).
+
+In the `pretrained_model_tuner.py` file, you'll find the code that defines both
```
Good catch! I got rid of that hyphenated one.

flippedcoder commented:
> * Adding `model_name` to the parameters makes it so you can see at a glance whether an experiment is alexnet or squeezenet.

Good point! I added that param to the stage creation and in the repo.

flippedcoder commented:
> * Logging each epoch using dvclive or using checkpoints enables tracking results per epoch, which currently are getting printed to the screen and then lost.

I'll do that! I'm trying to break myself out of thinking in just experiments. 😅

flippedcoder commented:
> * Saving both the training and validation metrics shows a more complete picture (it looks like only the training metrics are being saved now) and you can choose to view one or the other or both.

Done!

flippedcoder commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,352 @@
+
+With all of this in place, we can finally start running experiments to fine-tune
+the two models.
+
```
If you clone the repo, then nope, there's no need for the steps above. I can cut that and just push for them to clone the repo and I'll start doing that for other blog posts.

flippedcoder commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,352 @@
+
+We'll add a stage for the training step in this example. To do that, run:
+
+```dvc
````
1. Done!
2. Working on adding checkpoints.
3. I'll add the model as an output to the stage.

daavoo commented:
> Fun article idea!

I also like the idea very much. 

I would expect the target audience to be someone at least familiar with "finetuning" but that would like to know about how to do it the "DVC way" and my general feedback is based on that assumption, although some things might be also benefitial to other targets. 

Trying to view it from that target perspective, I would love to read something that feels easily reusable for my custom use case. Could be even a way to engage with people at the end of the post with stuff like: now you go and try it with your data, or try to improve our results by switching params, etc.

In that regard, I think that it could be good to dig a little bit in the data preparation phase, using as base the [Data Versioning Tutorial](https://dvc.org/doc/use-cases/versioning-data-and-model-files/tutorial) could be useful for adding references to a more detailed tutorial, where the experimentation part kind of feels outdated. 

In addition, I have no strong preferences regarding *ML Frameworks* but my general feeling is that using a higher level framework (i.e. keras) would reduce the amount of boilerplate code that could make it harder to differenciate the relevant "finetuning with DVC" parts from regular boilerplate.
 
>  It might make sense to touch base with @daavoo whose background is in computer vision.

 It's kind of tricky to know beforehand which experiments would be meaningful 😅  but I would try to share a few ideas/intuitions that might showcase the benefits of experimentation. 

For pretty much every idea, I think the following things would be helpful to support discussions:

- Add per-epoch and per split (train/val) metrics (could be done via DVCLive as mentioned by Dave)
- Add `training time` or `time per epoch` as a metric. This might be tricky for many reasons but even a naive measure would be useful.

Here are those ideas for experimentation "à la DVC":

### Adding more training data

**Depends on**: Having different dataset versions. Having `training_time` as metric.

**Showcases**: DVC data versioning.

**I would expect that**: extending the training set would improve perfomance. Even if that its not the case we might use the results to introduce another useful but less discussed topic: extending the validation set. 

### Comparing model architectures

**Depends on**: Having `model_name` as param. Having `training_time` as metric.

**Showcases**: `DVC params` 

**I would expect that**: 

Using a smaller version (lower `training_time`) of one family of architectures (i.e. `ResNet18`) would have worse validation accuracy than a bigger version of the same family (i.e. `ResNet50`). If that it's not the case might indicate that our problem/dataset is too simple (i.e. need to extend validation dataset to properly showcase this). 

Using a smaller model from newer families (i.e. `MobilenetV3`) would have better or similar accuracy than a bigger model from older families (i.e. `VGG19`)

**Note**: We are currently limiting the models I assume that due to torchvision not being very friendly about finetuning (i.e. can't easily reuse the same "add a new head" code for different architectures) I would either add a few if/elses for different architectures or switch to a framework where the code for adding a new head is more reusable.

### Comparing Finetuning strategies

We could use similar terminology to the one used in https://lightning-flash.readthedocs.io/en/latest/general/finetuning.html#finetune-strategies

**Depends on**: Having `feature_extract` as param. Having `training_time` as metric.

**Showcases**: `DVC params` 

**I would expect that**:  Training the entire model would result in higher `training_time`, for sure.; I'm not sure about `accuracy`. The theory says that, given a large enough dataset, training the entire model would have better perfomance than training only the head.


### Comparing Input Sizes / Data Augmentation

**Depends on**: Adapting the instantiation of `data_transforms` to be dynamic based on input `params` (I can share a code snippet for this). Having access to both `train` and `validation` metrics.

**Showcases**: `DVC params` 

**I would expect that**:  

Using smaller input size would result in lower `training_time` and  potentially lower `accuracy`.

Assuming we remove any data augmentation from the baseline, adding data augmentation (i.e. `RandomHorizontalFlip`) should improve `validation accuracy`.

---

There are probably other things that could be exposed in a similar way to `model_name`, like the Optimizer to be used (`optimizer_name`) however I think that the above ideas would be more understable for users less familiar with Deep Learning  and have more predictable results. 

Smaller vs bigger model, more data vs less data, "fancy new" vs "old school" architecture, etc. feels simpler than `momentum=0.9` vs `momentum=0.5`

flippedcoder commented:
> <img alt="Screen Shot 2021-08-05 at 5 52 59 PM" width="978" src="https://user-images.githubusercontent.com/3659196/128439529-d71df488-4f4e-45cd-93f0-1879fdcd10fa.png">
> 
> Something is wrong, does it work for you locally?
> 
> https://dvc-org-pre-defined-mod-kedssi.herokuapp.com/blog/pre-defined-models-exp is also broken it looks like ..

I'm having issues with the `sharp` package locally. Not sure if that's related to this though.

shcheklein commented:
@flippedcoder 

that's what I see in the build logs:

```
error Error processing Markdown file /tmp/build_6baa4287/content/blog/2021-08-24-pre-defined-models-exp.md:
      can not read a block mapping entry; a multiline key may not be an implicit key at line 6, column 16:
    descriptionLong: |
                   ^
       not finished source and transform nodes - 1.106s
```

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,369 @@
+description: |
+  You can work with pretrained models and fine-tune them with DVC experiments.
+descriptionLong: |
+  Running experiments to get the best tuning for a model can make it difficult to see which changes led to a better result. That's why we will be using DVC to track changes in the code and the data.
```
💅 please let's wrap it at 80 symbols please

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,369 @@
+
+Now that you have DVC initialized, this is a good place to commit your changes.
+
+```git
````
git -> dvc

shcheklein commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,369 @@
+can get the number of model features and the input size we need for fine-tuning.
+
+_You can find the original code and tutorial over at
+[this post](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html)._
```
rendering is broken

![Uploading Screen Shot 2021-08-10 at 3.59.26 PM.png…]()

flippedcoder commented:
@dberenbaum @shcheklein I added checkpoints to the project, updated the explanation of the yaml to match this, and updated the tables. Let me know if there's anything else you see that should be cleaned up or added!

dberenbaum commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,398 @@
+momentum: 0.017
+```
+
+Now run another experiment with `dvc exp run` and take a look at the metrics
````
The downside to adding checkpoints is that it changes the workflow. As you see in the table, `dvc exp run` will continue training additional epochs on top of your previous experiment. To start back at step 0, you will need to do `dvc exp run --reset`.

dberenbaum commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,398 @@
+several ways to do this with DVC:
+
+- Change the hyperparameter values directly in `params.yaml`
+- Update the values using the `--set-param` option on `dvc exp run`
```
Minor suggestion to introduce `-S` here also so that you can shorten all the commands below.

dberenbaum reviewed (changes requested):
Looks good, thanks @flippedcoder! I have a few more comments, mostly about either how to keep winnowing down the setup section or else addressing changes introduced by dvclive and checkpoints. IMO it's ready to go once those are fixed.

flippedcoder commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,398 @@
+
+We'll be fine-tuning the AlexNet model and the SqueezeNet model to classify
+images of bees and ants. You can find the project we're working with in
+[this repo](https://github.com/iterative/pretrained-model-demo).
```
Yeah, I like how that flows better. I got rid of the other link to the original project in favor of this.

flippedcoder commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,398 @@
+$ dvc exp run
+```
+
+This will execute the `pretrained_model_tuner.py` script and run for 5 epochs
````
Ope! I pushed my stuff to the repo after running my experiments. 😅 That _should_ be 5 now.

jorgeorpinel commented:
@dberenbaum @shcheklein @daavoo thanks for reviewing. It looks like @flippedcoder has addressed everything? Hoping for some approvals soon

dberenbaum commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,374 @@
+Now run another experiment with `dvc exp run` and take a look at the metrics
+with `dvc exp show`. Your table should look something like this now. Since we're
+using checkpoints, note that we continue training additional epochs on top of
+your previous experiment.
```
Do you think it's worth mentioning `--reset` here or saying that training from scratch will be shown below? Just a thought, but up to you.

dberenbaum commented on content/blog/2021-08-24-pre-defined-models-exp.md:
```diff
@@ -0,0 +1,374 @@
+┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━┩
+│ **workspace**               │ -        │   **13** │ **0.86066** │ **0.36708** │        **87.417** │ **0.025** │ **0.5**      │ **squeezenet** │ **2**           │ **8**          │ **2**          │
+│ **main**                    │ **10:27 AM** │    **1** │ **0.88525** │ **0.28284** │        **82.143** │ **0.01**  │ **0.09**     │ **squeezenet** │ **2**           │ **8**          │ **2**          │
+│ │ ╓ a771043 [exp-3d76f] │ 10:46 AM     │    1 │ 0.89754 │ 0.28754 │        82.447 │ 0.025 │ 0.5      │ squeezenet │ 2           │ 8          │ 2          │
```
Something happened with this table so that the columns aren't aligning.

dberenbaum commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,374 @@
+You can check out the details for the queues you have in place by looking at the
+experiments table with `dvc exp show`. You'll see something like this.
+
+```dvctable
````
`exp-7aac` in these last two tables looks off to me. It looks like you have squeezenet checkpoints that are running on top of alexnet checkpoints, which doesn't seem like the workflow you have described.

flippedcoder replied:
Yeah, not sure what I did there but it should be cleared up now!

dberenbaum commented:
There are a couple of table formatting issues I see. Once those look good, I'm happy to approve 🙂 .

dberenbaum commented on content/blog/2021-08-24-pre-defined-models-exp.md:
````diff
@@ -0,0 +1,374 @@
+each epoch using `dvclive` logging.
+
+```python
+if phase == 'train':
````
One more thought after taking a quick look at https://github.com/iterative/pretrained-model-demo/blob/132b6928ab11bbdb500e204c18e6491a03944b6b/pretrained_model_tuner.py#L84. The validation metrics are still being manually logged the old way and not captured by dvc. You could consider logging them with dvclive also like:

```python
            if phase == 'train':
                torch.save(model.state_dict(), "model.pt")

                dvclive.log('train_acc', epoch_acc.item())
                dvclive.log('train_loss', epoch_loss)
                dvclive.log('train_time', epoch_time_elapsed)

            if phase == 'val':
                dvclive.log('val_acc', epoch_acc.item())
                dvclive.log('val_loss', epoch_loss)
               
                dvclive.next_step()
```

Unfortunately, that does make your table even wider since you are adding more metrics. Another option might be to only log the validation metrics, since those are usually more helpful than training metrics, or to introduce the `--include-metrics`/`--include-params` flags to narrow the table.

flippedcoder replied:
I added those metrics to the table and I took out a few of the params we aren't adjusting.

dberenbaum replied:
Sounds good! Not sure if you are still making updates, but make sure to put `dvclive.next_step()` inside `if phase == 'val'` or else you will be making two steps for every epoch (once for train, once for val).

flippedcoder replied:
Oh! That makes sense. I wasn't quite sure about that. I'll make those updates here soon!

flippedcoder replied:
@dberenbaum Just pushed the update for this!

dberenbaum commented:
Final note: feel free to ping me again to take a final look. I might miss when the comments get resolved otherwise.

flippedcoder commented:
@dberenbaum I fixed all of the tables/workflow stuff!

dberenbaum commented:
Thanks, @flippedcoder! Did you see my last comment in https://github.com/iterative/dvc.org/pull/2656#discussion_r692150367?

dberenbaum commented:
Thanks again @flippedcoder! There's still some misalignments between the table headers (starting with the `val_loss` column) and other rows in some of the tables. Are you seeing the same?

flippedcoder commented:
@dberenbaum I've been fiddling with the table for a while and it seems like it's a combination of the formatting and the number of columns. It lines up a little better without the colors and bolding on rows, otherwise I didn't find anything else I could do about it.

dberenbaum commented:
@flippedcoder So that's the way it looks when you copy and paste from your terminal? Does the table look fine in your terminal?

IMO it's still worth cleaning up by manually adding/deleting spaces as needed so that it looks like the actual table output. I updated one of the tables in https://github.com/iterative/dvc.org/pull/2656/commits/7ba79bb29dc00a7f42fb8090a66ee429b43a46d5. Could you try to do the same for the tables below?

flippedcoder commented:
> So that's the way it looks when you copy and paste from your terminal? Does the table look fine in your terminal?

@dberenbaum It's weird because in my terminal it looks fine, but when I apply the formatting code it shifts things. I'll go through and manually update them then!

flippedcoder commented:
@dberenbaum The problem was with the table headers. It should be all cleared up now!

shcheklein commented:
There is a a broken link https://github.com/iterative/pretrained-model-demo (project is private - do we want to make it public before merge?)

Could we also try to minify static/uploads/images/2021-08-24/pretrained-models.png ? tinypng or some other similar service works.

jendefig commented:
> Could we also try to minify static/uploads/images/2021-08-24/pretrained-models.png ? tinypng or some other similar service works.

@shcheklein I just ran the image through tinypng a second time.  It shaved off another 2%.  I guess it's bigger than normal because it has all the images in it?  Roman had already compressed it.

flippedcoder commented:
> There is a a broken link https://github.com/iterative/pretrained-model-demo (project is private - do we want to make it public before merge?)

@shcheklein I just changed the repo to public and it looks like everything's good to go! Would you mind merging this?

shcheklein commented:
@flippedcoder done, merged 🎉 !
